schedule/scheduler.go

```diff
@@ -14,159 +18,301 @@ import (
+
+		scheduledJobComplete: make(chan struct{}, len(scheduledJobs)),
+		scheduledJobsRunning: len(scheduledJobs),
+
+		workerJobComplete: make(chan struct{}),
+		workerJobsRunning: 0,
 	}
-	return &Scheduler{
-		taskDelay: taskDelay,
+
+	// scheduled jobs added here will be started when Scheduler.Run is called.
+	for _, st := range scheduledJobs {
+		s.jobID++
+		st.id = s.jobID
+		st.log = log.With("id", st.id, "name", st.Name)
+		s.jobs[s.jobID] = st
 	}
+	return s
+}
+
+func NewSchedulerDaemon(mctx helpers.MetricsCtx, lc fx.Lifecycle) *Scheduler {
+	s := NewScheduler(0)
+	// Beware: starting things this way make the daemon unresponsive to api calls and kill signals..
+	/*
+		lc.Append(fx.Hook{
+			OnStart: s.Run,
+			OnStop:  nil,
+		})
+	*/
+	go func() {
+		if err := s.Run(mctx); err != nil {
+			log.Errorw("Scheduler Stopped", "error", err)
+		}
+	}()
+
+	return s
 }
 
 type Scheduler struct {
-	tasks     []TaskConfig
-	taskDelay time.Duration
+	jobs   map[JobID]*JobConfig
+	jobID  JobID
+	jobsMu sync.Mutex
+
+	jobDelay time.Duration
+
+	context context.Context
+
+	jobQueue chan *JobConfig
+
+	scheduledJobComplete chan struct{}
+	scheduledJobsRunning int
+
+	workerJobComplete chan struct{}
+	workerJobsRunning int
 }
 
-// Add add a task config to the scheduler. This must not be called after Run.
-func (s *Scheduler) Add(tc TaskConfig) {
-	s.tasks = append(s.tasks, tc)
+func (s *Scheduler) Submit(tc *JobConfig) JobID {
+	s.jobsMu.Lock()
+	s.jobID++
+	tc.id = s.jobID
+	s.jobsMu.Unlock()
+
+	s.jobQueue <- tc
+	return s.jobID
 }
 
-// Run starts running the scheduler and blocks until the context is done or
-// all tasks have run to completion.
+// Run starts running the scheduler and blocks until the context is done.
 func (s *Scheduler) Run(ctx context.Context) error {
-	if len(s.tasks) == 0 {
-		return xerrors.Errorf("no tasks to run")
-	}
+	log.Info("Starting Scheduler")
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
+	// used as context for jobs submitted, ensure they are canceled when context is canceled.
+	s.context = ctx
 
-	taskComplete := make(chan struct{}, len(s.tasks))
-	tasksRunning := len(s.tasks)
-
-	for _, tc := range s.tasks {
-		go func(tc TaskConfig) {
-			// Report task is complete when this goroutine exits
-			defer func() {
-				taskComplete <- struct{}{}
-			}()
-
-			// Attempt to get the task lock if specified
-			if tc.Locker != nil {
-				if err := tc.Locker.Lock(ctx); err != nil {
-					if errors.Is(err, storage.ErrLockNotAcquired) {
-						log.Infow("task not started: lock not acquired", "task", tc.Name)
-						return
-					}
-					log.Errorw("task not started: lock not acquired", "task", tc.Name, "error", err.Error())
-					return
-				}
-				defer func() {
-					if err := tc.Locker.Unlock(ctx); err != nil {
-						if !errors.Is(err, context.Canceled) {
-							log.Errorw("failed to unlock task", "task", tc.Name, "error", err.Error())
-						}
-					}
-				}()
+	for _, tc := range s.jobs {
```

Comment:
Guard access to s.jobs with mutex

Reply:
I have addressed the comment [here](https://github.com/filecoin-project/sentinel-visor/pull/414/files#r599600777) by removing the locks previously taken before this loop. They are not required since jobs cannot be access concurrently at this point. Thoughts?

Reply:
I'll need to re-review to follow the new logic.